lib/improver/wind_downscaling.py

```diff
@@ -0,0 +1,967 @@
+                    xpos = iii
+                elif mcc == mcube.coord(self.y_name):
+                    ypos = iii
+                elif mcc == mcube.coord(self.z_name):
+                    zpos = iii
+                elif mcc == mcube.coord(self.t_name):
+                    tpos = iii
+            except Exception:
+                pass
+        return xpos, ypos, zpos, tpos
+
+    def find_heightgrid(self, wind):
+        """Setup the height grid.
+
+        Setup the height grid either from the 1D or 3D height grid
+        that was supplied to the plugin or from the z-axis information
+        from the wind grid.
+
+        Parameters:
+        ----------
+        wind: 3D or 4D iris cube
+            representing the wind data.
+
+        Returns:
+        -------
+        hld: 1D or 3D np.array
+            representing the height grid.
+
+        """
+        if self.height_levels is None:
+            hld = wind.coord(self.z_name).points
+        else:
+            hld = iris.util.squeeze(self.height_levels)
+            if np.isnan(hld.data).any() or (hld.data == RMDI).any():
+                raise ValueError("height grid contains invalid points")
+            if hld.ndim == 3:
+                try:
+                    xap, yap, zap, _ = self.find_coord_order(hld)
+                    hld.transpose([yap, xap, zap])
+                except Exception:
+                    raise ValueError("height grid different from wind grid")
+            else:
+                try:
+                    hld = next(hld.slices([self.z_name]))
+                except Exception:
+                    raise ValueError("height z coordinate differs from wind z")
+            hld = hld.data
+        return hld
+
+    def check_wind_ancil(self, xwp, ywp):
+        """Check wind vs ancillary file grids.
+
+        Check if wind and ancillary files are on the same grid and if
+        they have the same ordering.
+
+        Parameters:
+        xwp: integer
+            representing the position of the x-axis in the wind cube
+        ywp: integer
+            representing the position of the y-axis of the wind cube
+
+        """
+        xap, yap, _, _ = self.find_coord_order(self.pp_oro)
+        if xwp - ywp != xap-yap:
+            if np.isnan(xap) or np.isnan(yap):
+                raise ValueError("ancillary grid different from wind grid")
+            else:
+                raise ValueError("xy-orientation: ancillary differ from wind")
+
+    def process(self, cube):
+        """Adjust the 4d wind field - cube - (x, y, z including times).
+
+        Parameters
+        ----------
+        cube - iris.cube.Cube
+            The wind cube to be operated upon. Should be wind speed on
+            height_levels for all desired forecast times.
+
+        Returns
+        -------
+        cube
+            The 4d wind field with roughness and height correction
+            applied in the same order as the input cube.
+
+        """
+        wind = cube
```

Comment:
I've renamed this as input_cube, to differeniate it from the resulting output_cube.